### ISSUE TITLE
Can't use lambdas with :keyword argument passing

### ISSUE THREAD

**Issue by cwebber** [hylang/hy#748]
Or: "Lambda hoisting strikes again!"

As a failure example:

```
(defn call-this [some-func]
  (some-func "yup"))

;; this works fine, as a "lambda one-liner"
(call-this :some-func (lambda [x] (print x)))

;; so does this, with a more complex lambda, as a positional argument
(call-this (lambda [x] (do (+ 1 2) (print x))))

;; but this doesn't, as a keyword argument
(call-this :some-func (lambda [x] (do (+ 1 2) (print x))))
```

hy2py results:

```
def call_this(some_func):
    return some_func('yup')
call_this(some_func=(lambda x: print(x)))

def _hy_anon_fn_2(x):
    (1 + 2)
    return print(x)
call_this(_hy_anon_fn_2)
call_this(some_func=_hy_anon_fn_3)
```

**Comment by paultag** [comment]
What exactly breaks?

(hy)[tag@cassiel:~][⌚ 06:06 PM] ♥  hy
hy 0.10.1 using CPython(default) 3.4.2 on Linux
=> (defn call-this [some-func]
...   (some-func "yup"))
=>
=> ;; this works fine, as a "lambda one-liner"
=> (call-this :some-func (lambda [x] (print x)))
yup
=>
=> ;; so does this, with a more complex lambda, as a positional argument
=> (call-this (lambda [x] (do (+ 1 2) (print x))))
yup
=>
=> ;; but this doesn't, as a keyword argument
=> (call-this :some-func (lambda [x] (do (+ 1 2) (print x))))
yup
=>



-- 
:wq

**Comment by cwebber** [comment]
It's going to sound strange, but try executing it as a file instead of in the interpreter.  It works in the interpreter as I showed above if you enter it line by line... I think because the last one inherits the name of the former anonymous one.

**Comment by cwebber** [comment]
@paultag In fact (separate bug?) you can see that it's running the "prior" lambda.  In the same session as above, try this:

```
=> (call-this :some-func (lambda [x] (do (print "what") (print x))))
yup
=>
```

Oops, it should have printed "what" too!

**Comment by cwebber** [comment]
Looks like @zackmdavis fixed this in #796.  Closing!
